### PR TITLE
Replace sbt dependency with zinc

### PR DESCRIPTION
**Pull request by sschaef**
Scala IDE so far did depend directly on sbt to build Scala sources. Because zinc has been extracted out of sbt into its own [repo](https://github.com/sbt/zinc), we finally got the chance to get rid of the sbt dependency by only relying on zinc.

This greatly simplifies uber-build (see the PR: https://github.com/scala-ide/uber-build/pull/87) but not so much the scala-ide build itself because packaging zinc in a way that Eclipse likes it is still a major act.

In the following a short description of the changes:

- `org.scala-ide.sbt.compiler.interface` has been renamed to `org.scala-ide.zinc.compiler.bridge`. This is because the sbt dependency `compiler-interface` has been renamed to `compiler-bridge` in zinc. However, there is a newly introduced dependency also called `compiler-interface` in zinc (but it has a different purpose than the one in sbt).
- `org.scala-ide.sbt.full.library` has been renamed to `org.scala-ide.zinc.library`
- `org.scala-ide.sbt.build` has been renamed to `org.scala-ide.zinc`
- The reason why `org.scala-ide.zinc` is still needed as an Eclipse bundle is that I didn't find another way to bundle zinc together with `org.scala-ide.sdt.core`. The following things had to be achieved: 
  - `compiler-bridge` depends on scalac and is therefore binary incompatible between minor versions. We need to build it from source. This is done at runtime by the IDE in `CompilerBridgeStore`.
  - In order to allow the runtime compilation in `CompilerBridgeStore` we need to ship the sources of `compiler-bridge` and all of its dependencies as binaries (the dependencies do not depend on scalac and therefore can be shipped as binaries).
  - `org.scala-ide.zinc.library` contains the said binaries. `org.scala-ide.zinc.compiler.bridge.source` contains the said sources. `org.scala-ide.zinc` bundles both as an Eclipse bundle.
  - The bundling of `org.scala-ide.zinc.library` and `org.scala-ide.zinc.compiler.bridge.source` can't be done in `org.scala-ide.sdt.core` because the maven build first checks if the bundle is available before the building happens. Only with `org.scala-ide.zinc` we could achieve that the bundle is available at the dependency checking phase of maven. This is also the reason why `build-all.sh` exists, whose purpose it is to make the zinc and Scala dependencies available to scala-ide. With a single maven build this can't be achieved, because the bundles are created in a maven phase that happens after the dependency-checking phase and therefore these bundles can't be made available to other bundles of scala-ide.
- The manifests need to depend on all zinc artifacts because these artifacts are not available through p2 update sites.

TODO:

- [x] Bug in zinc needs to be fixed: https://github.com/sbt/zinc/issues/169
- [x] Find out what to do with `ClassfileManager` in `SbtInputs`
- [x] Fix test failures
- [x] Look through the code once again to do last cleanups.

### CODE REVIEW

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/NameHashingVulnerabilityTest.scala [hunk 1]
```diff
@@ -61,26 +59,21 @@ object NameHashingVulnerabilityTest extends IProjectOperations {
...
   private def thenThereIsNoErrors(): Unit = {
     val errors = markersMessages(findProjectProblemMarkers(project, errorTypes: _*).toList)
     Assert.assertTrue("no error expected: " + errors.mkString(", "), errors.isEmpty)
   }
 
-  private def whenNameHashingIs(isOn: Boolean)(otherwise: => Unit): Unit = {
-    val nameHashingProperty = SettingConverterUtil.convertNameToProperty(ScalaPluginSettings.nameHashing.name)
-    project.storage.setValue(nameHashingProperty, isOn)
+  private def whenNameHashingIsOn(otherwise: => Unit): Unit = {
+    project.storage.setValue("nameHashing", true)
```
Comment (wpopielarski) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/NameHashingVulnerabilityTest.scala]: I understand that `nameHashing` has got one value `true`

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/CachingCompiler.scala [hunk 2]
```diff
@@ -23,29 +34,38 @@ class CachingCompiler private (cacheFile: File, sbtReporter: Reporter, log: Logg
...
+
+      override def definesClass(classpathEntry: File) = {
+        val dc = Locator(classpathEntry)
+        new DefinesClass() {
+          def apply(name: String) = dc.apply(name)
+        }
+      }
+    }
+//    val aMap = (f: File) => SbtUtils.m2o(in.analysisMap(f))
+//    val defClass = (f: File) => { val dc = Locator(f); (name: String) => dc.apply(name) }
```
Comment (wpopielarski) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/CachingCompiler.scala]: not sure these two, rather shouldn't stay

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/CachingCompiler.scala [hunk 3]
```diff
@@ -23,29 +34,38 @@ class CachingCompiler private (cacheFile: File, sbtReporter: Reporter, log: Logg
...
+      SbtUtils.m2o(in.progress), in.scalacOptions, in.javacOptions, previousAnalysis, previousSetup, lookup,
+      sbtReporter, in.order, skip = false, in.incOptions, extra = Nil)(log))
   }
 
-  private def cacheAndReturnLastAnalysis(compilationResult: IC.Result): Analysis = {
+  private def cacheAndReturnLastAnalysis(compilationResult: CompileResult): Analysis = {
     if (compilationResult.hasModified)
       MixedAnalyzingCompiler.staticCachedStore(cacheFile).set(compilationResult.analysis, compilationResult.setup)
-    compilationResult.analysis
+    // TODO is this cast safe here?
```
Comment (wpopielarski) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/CachingCompiler.scala]: good remark

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/JavaEclipseCompiler.scala [hunk 4]
```diff
@@ -23,7 +22,7 @@ class JavaEclipseCompiler(p: IProject, monitor: SubMonitor) extends JavaCompiler
 
   override def project = p
 
-  override def compile(sources: Array[File], classpath: Array[File], output: Output, options: Array[String], log: Logger): Unit = {
+  override def run(sources: Array[File], options: Array[String], reporter: Reporter, log: Logger): Boolean = {
```
Comment (wpopielarski) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/JavaEclipseCompiler.scala]: not sure about this but I think would be fine to know from method signature that argument is not used like
`override def run(sources: Array[File], notUsedOptions: Array[String], notUsedReporter: Reporter, notUsedLog: Logger): Boolean ` (I hope I didn't make a mistake here)

**Comment by wpopielarski** [comment]
looks promising 👍

**Comment by sschaef** [comment]
retest this please

**Comment by wpopielarski** [comment]
I looked at tests and seems like javac is not called by incremental compiler

**Comment by wpopielarski** [comment]
My suspicion is that your `SbtInputs.analysisMap` needs to be written like in my last PR so it must not return `Maybe.just(Analysis.Empty)` but just `Maybe.nothing[Analysis]`
Here is a code:
```scala
  def analysisMap(f: File): Maybe[Analysis] =
    if (f.isFile)
      Maybe.nothing[Analysis]
    else {
      val analysis = allProjects.collectFirst {
        case project if project.buildManager.buildManagerOf(f).nonEmpty =>
          project.buildManager.buildManagerOf(f).get.latestAnalysis(incOptions)
      }
      analysis.map { analysis =>
        Maybe.just(analysis)
      }.getOrElse(Maybe.nothing[Analysis])
    }
```

**Comment by wpopielarski** [comment]
I try to build it but I have problem with following p2 repo: http://download.eclipse.org/releases/luna/ btw isn't be more like http://download.eclipse.org/releases/neon/ ?

**Comment by sschaef** [comment]
The test failures are weird. On my system all of them pass.

Where do you have any problems with the luna p2 update site? Everything should point to neon.

**Comment by wpopielarski** [comment]
I cannot build it. Ends up with:
[ERROR] Internal error: java.lang.RuntimeException: Failed to load p2 repository with ID 'eclipse.luna' from location http://download.eclipse.org/releases/luna: Unable to read repository at http://download.eclipse.org/releases/luna. Unable to read repository at http://download.eclipse.org/releases/luna/201502271000/content.jar. Read timed out

**Comment by wpopielarski** [comment]
I see my mistake, forgot to run git fetch origin

**Comment by wpopielarski** [comment]
I have 8 failed tests in core on Win. Looking into....

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/SbtInputs.scala [hunk 5]
```diff
@@ -60,93 +63,94 @@ class SbtInputs(installation: IScalaInstallation,
...
+  def sources = sourceFiles.toArray
+
+  def output = new MultipleOutput {
+    private def sourceOutputFolders =
+      if (srcOutputs.nonEmpty) srcOutputs else project.sourceOutputFolders
+
+    override def outputGroups = sourceOutputFolders.map {
+      case (src, out) => new MultipleOutput.OutputGroup {
+        override def sourceDirectory = {
+          val loc = src.getLocation
```
Comment (wpopielarski) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/SbtInputs.scala]: Option with map? Just proposal

File: org.scala-ide.sdt.core/src/org/scalaide/util/internal/SbtUtils.scala [hunk 6]
```diff
@@ -2,10 +2,23 @@ package org.scalaide.util.internal
...
   def m2o[S](opt: Maybe[S]): Option[S] = if (opt.isEmpty) None else Some(opt.get)
 
+  def readCache(cacheFile: File): Option[(Analysis, MiniSetup)] =
+    FileBasedStore(cacheFile).get().map(_ match {
+      case (a: Analysis, i) => (a, i)
+      case (a, _) => throw new RuntimeException(s"Expected that sbt analysis for $cacheFile is of type ${classOf[Analysis]} but was ${a.getClass}.")
+    })
+
+  def readAnalysis(cacheFile: File, incOptions: => IncOptions): Analysis =
+    readCache(cacheFile).map(_._1).getOrElse(Analysis.empty(nameHashing = incOptions.nameHashing()))
```
Comment (wpopielarski) [inline org.scala-ide.sdt.core/src/org/scalaide/util/internal/SbtUtils.scala]: `nameHashing` is `true` only?

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/CachingCompiler.scala [hunk 7]
```diff
@@ -23,29 +34,38 @@ class CachingCompiler private (cacheFile: File, sbtReporter: Reporter, log: Logg
...
    */
-  def compile(in: SbtInputs, comps: Compilers[AnalyzingCompiler]): Analysis = {
-    val options = in.options; import options.{ options => scalacOptions, _ }
-    val aMap = (f: File) => SbtUtils.m2o(in.analysisMap(f))
-    val defClass = (f: File) => { val dc = Locator(f); (name: String) => dc.apply(name) }
-    val (previousAnalysis, previousSetup) = SbtUtils.m2o(IC.readCache(cacheFile))
+  def compile(in: SbtInputs, comps: Compilers): Analysis = {
+    val lookup = new PerClasspathEntryLookup {
+      override def analysis(classpathEntry: File) =
+        in.analysisMap(classpathEntry).asInstanceOf[xsbti.Maybe[CompileAnalysis]]
```
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/CachingCompiler.scala]: fixed

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/EclipseSbtBuildManager.scala [hunk 8]
```diff
@@ -186,7 +185,7 @@ class EclipseSbtBuildManager(val project: IScalaProject, settings: Settings, ana
 
   // take by-name argument because we need incOptions only when we have a cache miss
   override def latestAnalysis(incOptions: => IncOptions): Analysis =
```
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/EclipseSbtBuildManager.scala]: Unfortunately `CompileAnalysis` is just a marker interface, we can't do anything with it.

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/SbtInputs.scala (hunk 5 above)
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/SbtInputs.scala]: We can't return an Option here

File: org.scala-ide.sdt.core/src/org/scalaide/util/internal/SbtUtils.scala (hunk 6 above)
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/util/internal/SbtUtils.scala]: fixed

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/NameHashingVulnerabilityTest.scala [hunk 9]
```diff
@@ -61,26 +59,21 @@ object NameHashingVulnerabilityTest extends IProjectOperations {
 
 class NameHashingVulnerabilityTest extends IProjectOperations with IProjectHelpers {
   import NameHashingVulnerabilityTest._
-  private val On = true
-  private val Off = false
   private val errorTypes = Array(SdtConstants.ProblemMarkerId, IJavaModelMarker.JAVA_MODEL_PROBLEM_MARKER)
 
   @Test def shouldCorrectlyBuildProjectForAllPossibleSettingsOfNameHashingFlag(): Unit = {
```
Comment (sschaef) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/NameHashingVulnerabilityTest.scala]: you are right, I removed it

File: org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/CompilerBridgeStoreTest.scala [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+class CompilerBridgeStoreTest {
+
+  @Test
+  def platformCompilerBridgeWorks(): Unit = {
+    val store = ScalaPlugin().compilerBridgeStore
+    store.purgeCache()
+
+    Assert.assertTrue("successful compiler bridge compilation", store.compilerBridgeFor(platformInstallation)(null).isRight)
+    Assert.assertEquals("Zero hits and one miss", (0, 1), store.getStats)
```
Comment (sschaef) [inline org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/CompilerBridgeStoreTest.scala]: The first number is hits, the second one is misses. It is explained in Scaladoc of `getStats`. Too lazy to convert that to convert that to a case class.

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/JavaEclipseCompiler.scala (hunk 4 above)
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/JavaEclipseCompiler.scala]: fixed

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/CachingCompiler.scala (hunk 2 above)
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/CachingCompiler.scala]: fixed

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/CachingCompiler.scala (hunk 3 above)
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/builder/zinc/CachingCompiler.scala]: fixed

**Comment by sschaef** [comment]
I fixed all remaining TODOs and applied all review comments. Looks ready to merge now.

**Comment by wpopielarski** [comment]
if you are sure about `CompileAnalysis` and `Analysis` then LGTM :+1:

**Comment by dwijnand** [comment]
Please note that anything in `sbt.internal` can change in any way, so it's a risk to depend on them.

It's part of the effort of exposing a smaller api so that it's less painful and limiting to sbt or zinc to evolve.

/cc  @eed3si9n

**Comment by sschaef** [comment]
We always dependent on zinc internals, from this point of view nothing changed. If there are larger changes planned, we would welcome if you could drop us a message (or drop it on gitter, I'm following sbt-dev anyway)

---- On Wed, 21 Sep 2016 23:02:56 +0200 notifications@github.com wrote ----

Please note that anything in sbt.internal can change in any way, so it's a risk to depend on them.

It's part of the effort of exposing a smaller api so that it's less painful and limiting to sbt or zinc to evolve.

/cc @eed3si9n

**Comment by dwijnand** [comment]
Nothing in zinc wasn't internal, ie. all of sbt was public API.

Now things that are in sbt.internal are expressly internal, so caveat emptor.

Things might continue to change in the future, I'll try to remember to ping you in sbt-dev.

Note that you're depending on a milestone pre-release of zinc...

**Pull request merged**